### PR TITLE
[Feature] Only group member can write posts

### PR DESCRIPTION
**Pull request by glasses618**
Feature:
1. Users can join/leave groups.
2. Only group member can write posts.

How to test:
Run rails server `$rails s`, and open http://localhost:3000 with your browser. 
You can join/leave groups after login.
You can write posts in a group once you join the group.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,6 +1,11 @@
 Rails.application.routes.draw do
   devise_for :users
   resources :groups do
+    member do
+      post :join
+      post :leave
+    end
```
Comment (haopingfan) [inline config/routes.rb]: 可以新開一個 group_users_controller, 然後用這個 controller 的 create, destroy action 取代 groups controller 的 join, leave action. `resource :group_users, only: [:create, :destroy]` 這樣更符合 RESTful 設計精神

File: db/migrate/20180103023036_create_group_relationships.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class CreateGroupRelationships < ActiveRecord::Migration[5.1]
+  def change
+    create_table :group_relationships do |t|
```
Comment (kakas) [inline db/migrate/20180103023036_create_group_relationships.rb]: 感覺叫 `group_user` 比較清楚

File: config/routes.rb (hunk 1 above)
Comment (kakas) [inline config/routes.rb]: 這邊我覺得可以用

```rb
namespace :group do
  resources: :members, only: [:create, :destroy]
  # 或 resources: :group_users, only: [:create, :destroy]
end
```

Comment (HEROGWP) [reply config/routes.rb]: 不過原本的寫法比較直覺ＸＤ

File: app/controllers/groups_controller.rb [hunk 3]
```diff
@@ -40,6 +41,26 @@ def destroy
...
+  end
+
+  def leave
+    if current_user.member_of?(@group)
+      current_user.leave(@group)
+      flash[:notice] = '成功退出本討論版'
+    else
+      flash[:warning] = '你不是本討論版成員'
+    end
+    redirect_to group_path(@group)
```
Comment (HEROGWP) [inline app/controllers/groups_controller.rb]: 跟上面的if 空個行

File: app/controllers/groups_controller.rb [hunk 4]
```diff
@@ -40,6 +41,26 @@ def destroy
...
   end
 
+  def join
+    if current_user.member_of?(@group)
+      flash[:warning] = '你已經是本討論版成員了'
+    else
+      current_user.join(@group)
+      flash[:notice] = '成功加入本討論版'
+    end
+    redirect_to group_path(@group)
```
Comment (HEROGWP) [inline app/controllers/groups_controller.rb]: 跟上面的if 空個行

File: app/controllers/user/participated_groups_controller.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    end
+
+    redirect_to group_path(@group)
+  end
+
+  private
+
+  def should_have_group
+    @group = Group.find_by(id: params[:group_id])
+    redirect_to groups_path, alert: 'Record not found' if @group.nil?
```
Comment (kakas) [inline app/controllers/user/participated_groups_controller.rb]: `Group not found` ?

**Pull request merged**